firebase-common/src/main/java/com/google/firebase/emulators/EmulatorSettings.java

```diff
@@ -0,0 +1,104 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.emulators;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import com.google.firebase.components.Preconditions;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Settings that control which Firebase services should access a local emulator, rather than
+ * production.
+ *
+ * <p>TODO(samstern): Un-hide this once Firestore, Database, and Functions are implemented
+ *
+ * @see com.google.firebase.FirebaseApp#enableEmulators(EmulatorSettings)
+ * @hide
+ */
+public class EmulatorSettings {
+
+  /** Empty emulator settings to be used as an internal default */
+  public static EmulatorSettings getDefault() {
+    return new EmulatorSettings.Builder().build();
+  }
+
+  public static final class Builder {
+
+    private final Map<FirebaseEmulator, EmulatedServiceSettings> settingsMap = new HashMap<>();
+
+    /** Constructs an empty builder. */
+    public Builder() {}
+
+    /**
+     * Specify the emulator settings for a single service.
+     *
+     * @param emulator the emulated service.
+     * @param settings the emulator settings.
+     * @return the builder, for chaining.
+     */
+    @NonNull
+    public Builder addEmulatedService(
+        @NonNull FirebaseEmulator emulator, @NonNull EmulatedServiceSettings settings) {
+      Preconditions.checkState(
+          !settingsMap.containsKey(emulator),
+          "Cannot call addEmulatedService twice for " + emulator.toString());
+      this.settingsMap.put(emulator, settings);
+      return this;
+    }
+
+    @NonNull
+    public EmulatorSettings build() {
+      return new EmulatorSettings(new HashMap<>(settingsMap));
+    }
+  }
+
+  private final AtomicBoolean accessed = new AtomicBoolean(false);
+  private final Map<FirebaseEmulator, EmulatedServiceSettings> settingsMap;
+
+  private EmulatorSettings(@NonNull Map<FirebaseEmulator, EmulatedServiceSettings> settingsMap) {
+    this.settingsMap = Collections.unmodifiableMap(settingsMap);
+  }
+
+  /**
+   * Determine if any Firebase SDK has already accessed the emulator settings. When true, attempting
+   * to change the settings should throw an error.
+   *
+   * @hide
+   */
+  public boolean isAccessed() {
+    return accessed.get();
+  }
+
+  /**
+   * Fetch the emulation settings for a single Firebase service. Once this method has been called
+   * {@link #isAccessed()} will return true.
+   *
+   * @hide
+   */
+  @Nullable
+  public EmulatedServiceSettings getServiceSettings(@NonNull FirebaseEmulator emulator) {
+    accessed.set(true);
```

Comment:
The class as implemented appears immutable. It doesn't seem possible to call   `addEmulatedService` and modify an existing instance, so it is not immediately obvious why we need the "accessed" check. 

Would it make more sense to perform this check in `FirebaseApp` and enforce that `enableEmulators` can only be called once? Along the same line, could `enableEmulators` take `FirebaseEmulator emulator,  EmulatedServiceSettings settings` directly, which would remove the need for `EmulatorSettings` API.

Reply:
It's not about preventing someone calling `enableEmulators` twice, it's about someone calling it *after* the Database/Firestore/etc SDK has already consulted `FirebaseApp.getEmulatorSettings()`.

It's sort of like `firestore.settings({ })` where you can call it all you want but once the SDK "boots up" we throw if you call it.

As for changing the API to be repeated calls to `enableEmulators`:
  1. That would be a pretty significant departure from the API approved by the council.
  2. The idea here is that developers should have a *single* place where they control all the emulator settings in their app.  So I am personally in favor of a single call to `enableEmulators()` with all of the service settings in one object.

Reply:
Okay, understood. I think we should still try to simplify the code flow, which would make it more obvious to code reviewers like me. Can we make this explicit in Firebase App (by adding an internal `freezeSettings()` or something)? ideally, trying to understand this should not overheat my brain, but it is also very hot today.

Reply:
Yes that's a good idea.

Reply:
Added a hidden `freeze()` method on `EmulatorSettings` and made it the responsibility of `FirebaseApp` to call it.

Added a warning if somehow someone is getting service settings from a non-frozen `EmulatorSettings`, which shouldn't be possible.

Reply:
Hm, that's interesting. What's become more apparent to me with your latest changes is the observation that `EmulatorSettings` are not really concerned about being frozen, it's `FirebaseApp` that both `freezes` and checks for `frozen` on it.

This kind of results in the situation where the `frozen` logic is spread out among both `FirebaseApp` and `EmulatorSettings`. Would it make sense to make `EmulatorSettings` just a pure immutable pojo and have `FirebaseApp` store the `frozen` bit since it's the only one who uses it and has full control of the value?

Reply:
That's cool with me!  I agree that's where it should live, I just thought maybe people didn't want that kind of state on `FirebaseApp`.  I'll move it.

Reply:
Moved boolean state to `FirebaseApp`